### PR TITLE
feat: initial draft of spec explaining peering, clearing and settling

### PR DESCRIPTION
**Pull request by adrianhopebailie**

### CODE REVIEW

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+The Interledger network is a graph of nodes (connectors) that have peered with one another by establishing a means of exchanging ILP packets and a means of paying one another for the successful forwarding and delivery of the packets.
+
+Further details on the ILP Protocol can be found in the [specification](../0027-interledger-protocol-4/0027-interledger-protocol-4.md).
+
+# Accounts and Balances
+
+The edge between any two nodes (peers) is a communication link (for exchanging ILP packets and other data), and an account held between the peers (the Interledger account). The Interledger account has a balance denominated in a mutually agreed upon asset (e.g. USD) at an agreed upon scale (e.g. 2). The balance on the account is **the net total of the amounts on any packets successfully routed between the peers**.
+
+The balance on the Interledger account can only change as a result of two events:
+ 1. The successful routing of an ILP packet between the two peers
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: Might be worth mentioning here what successful means, or mentioning that it'll be explained below

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+draft: 1
+---
+
+The Interledger network is a graph of nodes (connectors) that have peered with one another by establishing a means of exchanging ILP packets and a means of paying one another for the successful forwarding and delivery of the packets.
+
+Further details on the ILP Protocol can be found in the [specification](../0027-interledger-protocol-4/0027-interledger-protocol-4.md).
+
+# Accounts and Balances
+
+The edge between any two nodes (peers) is a communication link (for exchanging ILP packets and other data), and an account held between the peers (the Interledger account). The Interledger account has a balance denominated in a mutually agreed upon asset (e.g. USD) at an agreed upon scale (e.g. 2). The balance on the account is **the net total of the amounts on any packets successfully routed between the peers**.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: Might want to explain asset scale a bit more

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+The balance on the Interledger account can only change as a result of two events:
+ 1. The successful routing of an ILP packet between the two peers
+ 1. A payment made between the peers on the underlying payment network they have agreed to use to settle the Interledger account
+
+ This document provides further details on the when these events occur andhow they are handled by the peers. 
+
+> **Example : Alice and Bob**
+
+> - Alice and Bob are peers on the network.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: Nitpick: Are they peers before they set up their peering relationship? Before that they are just nodes on the Interledger network

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+When a connector sends an ILP Prepare packet to a peer, and receives a valid ILP Fulfill response, before the expiry of the ILP Prepare, their balance with that peer **decreases**. 
+
+When a connector receives an ILP Prepare packet and returns a valid ILP Fulfill response, before the expiry of the ILP Prepare, their balance with that peer **increases**.
+
+One peer's balance should always be the additive inverse (negation) of the other peer's balance. I.e. If one peer has a balance of 10 the other peer should have a balance of -10.
+
+In financial accounting terms this could be viewed as a revenue/expense account where credits to this account are balanced by a debit to a corresponding "Accounts Receivable" asset account and debits to this account are balanced by a credit to a corresponding "Accounts Payable" liability account.
+
+> **Example : Alice and Bob**
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: This shouldn't have the same title as above. Maybe say this the accounting or balances for the Alice and Bob example

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+> 1. Alice passes an ILP Prepare to Bob with an amount of 6.
+> 1. Bob replies, before the request expires, with a valid ILP Fulfill.
+> 1. Bob passes an ILP Prepare to Alice with an amount of 2.
+> 1. Alice is unable to route the packet and replies with an ILP Reject.
+> 1. Bob passes an ILP Prepare to Alice with an amount of 20.
+> 1. Alice replies, before the request expires, with a valid ILP Fulfill.
+
+> The balance changes will be as follows:
+
+> - After Step 2:
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: I think it would be clearer to say what the balances are in the list of steps above

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+>   - Alice's balance is -6 satoshis
+>   - Bob's balance is 6 satoshis
+> - After Step 4 there are no changes to the balance as the packet was not delivered.
+> - After Step 6:
+>   - Alice's balance is 14 satoshis
+>   - Bob's balance is -14 satoshis
+
+# Clearing
+
+The traditional definition of clearing a payment, is the process by which the two parties establish their net positions and reconcile any differences.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: No comma needed

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+In the Interledger case this would involve the two peers exchanging data on what their current balance is and, if there is a discrepency, resolving this.
+
+Notably, clearing is not part of the Interledger standard, as the means by which two peers perform these operations is a bilateral concern (i.e. has no impact on others in the network).
+
+Current implementations of ILP do not have a standardized clearing process, as most peers settle very frequently using networks on which this is possible such as distributed ledgers which support payment channels, reducing the risk of discrepencies developing in their account balances.
+
+However, there is an edge case where the peer's balances may be out of sync demonstrated in this example.
+
+> **Example : Alice and Bob**
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: Again, what is different about this case?

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+> The balance changes will be as follows:
+
+> - After Step 2:
+>   - Alice's balance is 20 satoshis
+>   - Bob's balance is 0 satoshis
+
+> Clearly, Bob and Alice need to reconcile this situation. The specific mechanism they use and recourse they take to resolve this difference is outside the scope of the Interledger protocol. 
+
+It is important to note that this is an extremely unlikely scenario as telescoping expiry timeouts between peers should reduce the risk of this to almost zero.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: It may happen but nodes control how much time they subtract from the expiry time before forwarding it (worth explaining the telescoping bit and emphasizing that it's under their control)

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+Further, this is a single packet, and a key principal of the Interledger network is to break payments into very small packets so that in the unlikely event that this does occur the losses incurred by the two parties are negligible.
+
+Nonetheless, it is likely that future bilateral protocols (like BTP) will define more detailed processes for reconciliation and clearing.
+
+# Settlement
+
+It is unlikely that the flow of packets between two peers will always net out such that the balance continuously tends toward zero. In most cases flows will be predominantly in one direction and there will be a need for the peers to reduce them back to 0.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: It'll be reduced or raised to zero, depending on your perspective

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+This avoids a situation where one peer owes the other a large sum of money and is then unable or unwilling to pay (settlement risk). Connectors should define a maximum Interledger account balance to manage the settlement risk they expose themselves to with their peers.
+
+Further, the principal of the Interledger network is to settle as often as is practically possible and financially viable such that this risk is very low.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: I'm not sure that's a principle (also typo with that word) of the network, but more like a mechanism we've found very helpful especially between peers that do not have a lot of bilateral credit/trust. Over time that could change

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+In the case of peers that use payment channels to settle this can be done for every packet as the cost of settlement is almost zero (the cost of exchanging claims on the channel). This reduces the settlement risk between peers down to the value of only those packets that are not yet fulfilled, rejected or expired (i.e. packets that are in-flight).
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: "exchanging claims" -> "exchanging claims or updates" (XRP uses the term claims whereas other networks describe them differently)

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+Connectors will configure their own business rules regarding when to settle, based on how much they trust their peers and the costs and speed of settlement on the underlying network.
+
+For most implementations, this configuration will consist of, at a minimum, a maximum balance, and a settlement threshold. 
+
+## Maximum Balance
+
+When a connector receives an incoming packet from a peer it will ensure that the amount of the packet, added to the current balance, does not exceed the maximum balance. If it does it MUST reject he packet with a T04 error.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: Include the error name

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+ 1. The settlement threshold is lower than the additive inverse (negation) of the maximum balance at the peer, or
+ 1. The sending peer is unable to settle the account fast enough to keep up with the total amount in the packets being sent.
+
+In the former case the peer may choose to perform a settlement, even though it has not reached its settlement threshold so that the balance on the Interledger account at the peer is reduced below the maximum. 
+
+In the latter case, the sending peer may need to throttle back on the packets it sends to the upstream peer or they will need to make an alternatiuve settlement arrangement that can accomodate the volume.
+
+## Settlement Threshold
+
+When a connector receives a successful response from a peer it will evaluate if the amount of the packet, subtracted from the current balance, is less than the settlement threshold. If so, the connector should perform a settlement, following which the balance on the Interledger account is adjusted up by the settlement amount.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: This is a bit hard to follow so it might be good to add `(see below for an example)`

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+> 1. Alice replies with a valid ILP Fulfill. Alice's balance is now 19. Bob's balance is -19.
+> 1. Bob's balance is now past his settlement threshold of -15 so he makes a 19 satoshi Lightning payment to Alice and notifies her that he has settled the account. Bob's is now 0.
+> 1. Alice receives the payment and adjusts her balance to 0.
+> 1. Alice passes an ILP Prepare to Bob with an amount of 2.
+> 1. Bob replies with a valid ILP Fulfill. Bob's balance is now 2. Alice's balance is -2.
+> 1. Alice passes an ILP Prepare to Bob with an amount of 5.
+> 1. Bob rejects this with a T04 error code because this would take his balance over 5.
+
+> Note how, because Bob has used very conservative limits on his maximum balance, and Alice has used a larger number for her settlement threshold, a situation will regularly arise where Bob reject packets from Alice because she is going to exceed Bob's maximum balance but she never hits her settlement threshold. While Alice can overcome this by settling whenever Bob rejects a packet with a T04 error a better solution would be for Alice to configure her threshold to be within Bob's maximum balance.
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: It may also be worth mentioning using F08 errors to indicate the maximum allowable packet size. That can help tell senders how much they should send in each packet, which can help avoid this situation. Or we could leave this out and just suggest that these two should raise their limits with one another

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+Further, the principal of the Interledger network is to settle as often as is practically possible and financially viable such that this risk is very low.
+
+In the case of peers that use payment channels to settle this can be done for every packet as the cost of settlement is almost zero (the cost of exchanging claims on the channel). This reduces the settlement risk between peers down to the value of only those packets that are not yet fulfilled, rejected or expired (i.e. packets that are in-flight).
+
+Connectors will configure their own business rules regarding when to settle, based on how much they trust their peers and the costs and speed of settlement on the underlying network.
+
+For most implementations, this configuration will consist of, at a minimum, a maximum balance, and a settlement threshold. 
+
+## Maximum Balance
```
Comment (emschwartz) [inline 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: You should explain why there's the maximum balance. If you post-pay each packet, then the peer could theoretically steal that much money from you by not paying up. If it's pre-paid, the maximum balance would probably be the total that the peer has pre-paid

Comment (adrianhopebailie) [reply 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: That's a good point. I may add a section on different settlement models.

Comment (emschwartz) [reply 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: You could also include in that how you might set it up with a slower settlement method.
One really important point to get across is how the choice of settlement method, frequency, and credit limits is completely independent of what everyone else in the network does and how you are insulated from other people's choices.

File: 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md (hunk 14 above)
Comment (adrianhopebailie) [reply 0032-peering-clearing-settlement/0032-peering-clearing-settlement.md]: I think we need another whole doc explaining throughput/bandwidth
